### PR TITLE
Fix for #2750 and #2783

### PR DESCRIPTION
**Pull request by kajojify**
Continuation of https://github.com/mitmproxy/mitmproxy/pull/2761 
I added tests with mocked `open` function. I hope it is the right way to test such things. 

Now, when you press `w`, `e` or `b` and input invalid file path, you will get something like:

image

### CODE REVIEW

File: test/mitmproxy/addons/test_export.py [hunk 1]
```diff
@@ -93,6 +93,17 @@ def test_export(tmpdir):
         assert qr(f)
         os.unlink(f)
 
+        with mock.patch("mitmproxy.addons.export.open") as m:
+            m.side_effect = [PermissionError("Permission denied"),
+                             IsADirectoryError("Is a directory"),
+                             FileNotFoundError("No such file or directory")]
+            for i in range(3):
```
Comment (Kriechi) [inline test/mitmproxy/addons/test_export.py]: I would prefer looping through each error individually. You could make use of a new test function and `pytest.parametrize` to set the side effect (exception), and also the expected log message.

File: mitmproxy/addons/cut.py [hunk 2]
```diff
@@ -88,26 +88,29 @@ def save(
...
+            else:
+                with open(path, "a" if append else "w", newline='', encoding="utf8") as fp:
+                    writer = csv.writer(fp)
+                    for f in flows:
+                        vals = [extract(c, f) for c in cuts]
+                        writer.writerow(
+                            [strutils.always_str(x) or "" for x in vals]  # type: ignore
+                        )
+                ctx.log.alert("Saved %s cuts over %d flows as CSV." % (len(cuts), len(flows)))
+        except (PermissionError, IsADirectoryError, FileNotFoundError) as e:
```
Comment (mhils) [inline mitmproxy/addons/cut.py]: FileNotFound does not make sense here, but how about we just say `except IOError`?

File: test/mitmproxy/addons/test_export.py (hunk 1 above)
Comment (kajojify) [reply test/mitmproxy/addons/test_export.py]: Thanks for the advice :) 
I just followed this comment https://github.com/mitmproxy/mitmproxy/pull/2761#issuecomment-355808629
I'll update this PR.

File: mitmproxy/addons/cut.py (hunk 2 above)
Comment (kajojify) [reply mitmproxy/addons/cut.py]: @mhils I think `FileNotFound` makes sense :)
We can get this exception, when inputting `/ome/file.txt` as a filepath. 
I was thinking about `OSError`, but `IOError` looks more reasonable, thanks.

File: test/mitmproxy/addons/test_cut.py [hunk 3]
```diff
@@ -111,6 +111,36 @@ def test_cut_save(tmpdir):
...
+            m.side_effect = [PermissionError("Permission denied"),
+                             IsADirectoryError("Is a directory"),
+                             FileNotFoundError("No such file or directory")]
+            for i in range(3):
+                tctx.command(c.save, "@all", "request.method", f)
+            assert tctx.master.has_log("Permission denied", level="error")
+            assert tctx.master.has_log("Is a directory", level="error")
+            assert tctx.master.has_log("No such file or directory",
+                                       level="error")
+
```
Comment (mhils) [inline test/mitmproxy/addons/test_cut.py]: Doesn't the test below replace the stuff above?

File: test/mitmproxy/addons/test_export.py [hunk 4]
```diff
@@ -94,6 +94,21 @@ def test_export(tmpdir):
         os.unlink(f)
 
 
+@pytest.mark.parametrize("exception, log_message", [
+    (PermissionError, "Permission denied"),
+    (IsADirectoryError, "Is a directory"),
+    (FileNotFoundError, "No such file or directory")
```
Comment (mhils) [inline test/mitmproxy/addons/test_export.py]: Man, those FileNotFoundErrors when writing new files... 😄

Comment (mhils) [reply test/mitmproxy/addons/test_export.py]: Ok, nevermind... https://mitmproxy.slack.com/archives/C060SGTJ6/p1516711868000193 🙈 😄

File: test/mitmproxy/addons/test_cut.py (hunk 3 above)
Comment (kajojify) [reply test/mitmproxy/addons/test_cut.py]: \0/
Forget to get rid of it. Sorry.

**Comment by mhils** [comment]
👍

**Pull request merged**